experimental/hip/native_executable.c

```diff
@@ -0,0 +1,309 @@
+  // Abstract resource used for injecting reference counting and vtable;
+  // must be at offset 0.
+  iree_hal_resource_t resource;
+
+  iree_allocator_t host_allocator;
+
+  const iree_hal_hip_dynamic_symbols_t* symbols;
+
+  // The loaded HIP module.
+  hipModule_t hip_module;
+
+  iree_host_size_t entry_point_count;
+  // The list of entry point data pointers, pointing to trailing inline
+  // allocation after the end of this struct.
+  iree_hal_hip_kernel_params_t entry_points[];
+} iree_hal_hip_native_executable_t;
+// + Additional inline allocation for holding entry point information.
+
+static const iree_hal_executable_vtable_t iree_hal_hip_native_executable_vtable;
+
+static iree_hal_hip_native_executable_t* iree_hal_hip_native_executable_cast(
+    iree_hal_executable_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_native_executable_vtable);
+  return (iree_hal_hip_native_executable_t*)base_value;
+}
+
+// Verifies the structure of the flatbuffer so that we can avoid doing so during
+// runtime.
+//
+// There are still some conditions we must be aware of (such as omitted names on
+// functions with internal linkage), however we shouldn't need to bounds check
+// anything within the flatbuffer after this succeeds.
+static iree_status_t iree_hal_hip_native_executable_flatbuffer_verify(
+    iree_const_byte_span_t flatbuffer_data) {
+  if (!flatbuffer_data.data) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "flatbuffer data is not present");
+  }
+
+  // Run flatcc generated verification. This ensures all pointers are in-bounds
+  // and that we can safely walk the file, but not that the actual contents of
+  // the flatbuffer meet our expectations.
+  int verify_ret = iree_hal_hip_ExecutableDef_verify_as_root(
+      flatbuffer_data.data, flatbuffer_data.data_length);
+  if (verify_ret != flatcc_verify_ok) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "flatbuffer verification failed: %s",
+                            flatcc_verify_error_string(verify_ret));
+  }
+
+  iree_hal_hip_ExecutableDef_table_t executable_def =
+      iree_hal_hip_ExecutableDef_as_root(flatbuffer_data.data);
+
+  flatbuffers_string_vec_t entry_points_vec =
+      iree_hal_hip_ExecutableDef_entry_points_get(executable_def);
+  size_t entry_point_count = flatbuffers_string_vec_len(entry_points_vec);
+  for (size_t i = 0; i < entry_point_count; ++i) {
+    if (flatbuffers_string_len(
+            flatbuffers_string_vec_at(entry_points_vec, i)) == 0) {
+      return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "executable entry point %zu has no name", i);
+    }
+  }
+
+  iree_hal_hip_BlockSizeDef_vec_t block_sizes_vec =
+      iree_hal_hip_ExecutableDef_block_sizes_get(executable_def);
+  size_t block_size_count = iree_hal_hip_BlockSizeDef_vec_len(block_sizes_vec);
+  if (block_size_count == 0) {
```

Comment:
I think we should have an initial check to make sure `entry_point_count` is not zero. And then we can remove this check and rely on the following `entry_point_count != block_size_count` check later.